bin/generate.dart

```diff
@@ -353,23 +386,87 @@ class DiagramGenerator {
+    <String>['flutter', 'devices', '--machine'],
+    printOutput: false,
+  );
+
+  final Map<String, Map<String, String>> devices =
+      <String, Map<String, String>>{};
+  final List<dynamic> devicesJson = jsonDecode(result.stdout) as List<dynamic>;
+  for (final Map<String, dynamic> entry
+      in devicesJson.cast<Map<String, dynamic>>()) {
+    devices[(entry['name'] as String).toLowerCase()] = <String, String>{
+      'id': entry['id'] as String,
+      'targetPlatform': entry['targetPlatform'] as String
+    };
+  }
+  return devices;
+}
+
+String getDeviceList(Map<String, Map<String, String>> devices) {
+  final List<String> output = <String>[];
+  for (final String key in devices.keys) {
+    final String id = devices[key]!['id']!;
+    final String platform = devices[key]!['targetPlatform']!;
+    output.add('$key : $id ${id != platform ? ' ($platform)' : ''}');
+  }
+  return output.join('\n');
+}
+
+// This will default to macOS on macOS, linux on linux, and the first attached
+// Android device on Windows.
+String getDefaultDevice(Map<String, Map<String, String>> devices) {
+  if (devices.isEmpty) {
+    return '';
+  }
+  final String platform = Platform.operatingSystem.toLowerCase();
+  if (devices.keys.contains(platform)) {
+    if (platform != 'windows') {
+      // We don't yet support Windows as a device, since the process requires
+      // some Unix tools still (ffmpeg, optipng, etc.).
+      return devices[platform]!['id']!;
+    }
+  }
+  for (final String key in devices.keys) {
+    if (devices[key]!['targetPlatform']!.startsWith('android')) {
+      // If we're not on a supported platform, return the first Android device
+      // found, if any.
+      return devices[key]!['id']!;
+    }
+  }
+  return '';
+}
+
 Future<void> main(List<String> arguments) async {
+  final Map<String, Map<String, String>> devices = await listAvailableDevices();
+
   final ArgParser parser = ArgParser();
   parser.addFlag('help', help: 'Print help.');
-  parser.addFlag('keep-tmp', help: "Don't cleanup after a run (don't remove temporary directory).");
+  parser.addFlag('keep-tmp',
+      help: "Don't cleanup after a run (don't remove temporary directory).");
   parser.addOption('tmpdir',
-      abbr: 't', help: 'Specify a temporary directory to use (implies --keep-tmp)');
+      abbr: 't',
+      help: 'Specify a temporary directory to use (implies --keep-tmp)');
   parser.addOption('device-id',
-      abbr: 'd', help: 'Specify a device to use for generating the diagrams', defaultsTo: 'linux');
+      abbr: 'd',
+      help: 'Specify a device ID to use for generating the diagrams. Defaults '
+          'to the host platform that the script is run on, if that platform is '
+          'supported, or an attached device if not. Default on this machine is '
+          '${getDefaultDevice(devices)}. Available devices are:\n${getDeviceList(devices)}\n',
+      defaultsTo: getDefaultDevice(devices));
```

Comment:
no need to say what the default is in the help string, args prints it out regardless